src/python/pants/engine/selectors.py

```diff
@@ -16,13 +16,36 @@ class Get:
   """Experimental synchronous generator API.
 
   May be called equivalently as either:
-    # verbose form: Get(product_type, subject_declared_type, subject)
-    # shorthand form: Get(product_type, subject_type(subject))
+    # verbose form: Get(product, subject_declared_type, subject)
+    # shorthand form: Get(product, subject_declared_type(<constructor args for subject>))
   """
   product: Type
   subject_declared_type: Type
   subject: Any
 
+  def __await__(self) -> Generator[Any, Any, Any]:
+    """Allow a Get to be `await`ed within an `async` method, returning a strongly-typed result.
+
+    The `yield`ed value `self` is interpreted by the engine within `extern_generator_send()` in
+    `native.py`. This class will yield a single Get instance, which is converted into
+    `PyGeneratorResponse::Get` from `externs.rs` via the python `cffi` library and the rust
+    `cbindgen` crate.
+
+    This is how this method is eventually called:
+    - When the engine calls an `async def` method decorated with `@rule`, an instance of
+      `types.CoroutineType` is created.
+    - The engine will call `.send(None)` on the coroutine, which will either:
+      - raise StopIteration with a value (if the coroutine `return`s), or
+      - return a `Get` instance to the engine (if the rule instead called `await Get(...)`).
+    - The engine will fulfill the `Get` request to produce `x`, then call `.send(x)` and repeat the
+      above until StopIteration.
+
+    See more information about implementing this method at
+    https://www.python.org/dev/peps/pep-0492/#await-expression.
+    """
```

Comment:
Awesome! Thank you for explaining this!

Reply:
It took me a while!!! I'm not going to just let that time be wasted!